src/main.rs

```diff
@@ -33,98 +34,110 @@ mod template;
 type Feature<'a> = &'a str;
 type PackageName<'a> = &'a str;
 type RootFeature<'a> = (PackageName<'a>, Feature<'a>);
+type Result<T> = std::result::Result<T, Error>;
 
 const VERSION_ATTRIBUTE_NAME: &str = "cargo2nixVersion";
 
 fn main() {
     let args: Vec<String> = std::env::args().collect();
     let args: Vec<&str> = args.iter().map(AsRef::as_ref).collect();
+    if let Err(err) = try_main(&args) {
+        eprint!(colorify!(red_bold: "error: "));
+        eprintln!("{}", err);
+        std::process::exit(1);
+    }
+}
 
+fn try_main(args: &[&str]) -> Result<()> {
     match &args[1..] {
         ["--stdout"] | ["-s"] => generate_cargo_nix(io::stdout().lock()),
         ["--file"] | ["-f"] => write_to_file("Cargo.nix"),
         ["--file", file] | ["-f", file] => write_to_file(file),
         ["--help"] | ["-h"] => print_help(),
-        ["--version"] | ["-v"] => println!("{}", version()),
+        ["--version"] | ["-v"] => {
+            println!("{}", version());
+            Ok(())
+        }
         [] => print_help(),
         _ => {
             println!("Invalid arguments: {:?}", &args[1..]);
             println!("\nTry again, with help: \n");
-            print_help();
+            print_help()
         }
     }
 }
 
 fn version() -> Version {
-    Version::parse(env!("CARGO_PKG_VERSION")).expect("parse CARGO_PKG_VERSION")
+    // Since `CARGO_PKG_VERSION` is provided by Cargo itself, which uses the same `semver` crate to
+    // parse version strings, the `unwrap()` below should never fail.
+    Version::parse(env!("CARGO_PKG_VERSION")).unwrap()
```

Comment:
I think `unwrap()` and `expect()` both are fundamentally the same thing, with only difference is that the latter takes a user-provided string (which is a good documentation as well). 

Also, both are a form of _assert_; so if your assertion is that `unwrap()` should never fail, then _that_ assertion doesn't change if you use `expect()` instead (I mean you can still make the same assertion). The latter, however, has one added advantage: IF your assertion fails, you'll get a better error message. :-)

If that makes sense, I'd suggest to revert this change. :-)

Reply:
That makes sense in general sure! 👍 But the only way that this would fail is if Cargo itself is broken, at which point, you would have bigger problems than a stock panic message. I feel that the addition of the message is redundant here for this reason.

Reply:
The meaning of *assertion* is that the programmer thinks something is impossible. `unwrap()` and `expect()` both convey that. So if and when that assumption turned out to be wrong, then of course you have a _bigger_ program, irrespective of whichever assertion form you use, but then you'll not have a custom message to quickly narrow down the issue if you use `unwrap()`. That's my opinion.

Also, if a programmer thinks something is not impossible, then he/she should not use `unwrap()` and `expect()` instead should add the possible _error_ handling scenario. I mean, `expect()` is not for the scenario where you _expect_ an error.  It's the opposite, exactly like `unwrap()`.

Reply:
I understand, but the point that I'm trying to make is that if `env!("CARGO_PKG_VERSION")` fails, then your application is probably not going to compile at all. The `expect()` message is useful at run-time when debugging broken run-time behavior, sure, but `env!()` is a compile-time invocation. If Cargo is unable to pass in the value of `CARGO_PKG_VERSION` into `rustc` ([source](https://github.com/rust-lang/cargo/blob/d6fa260d1aa74521c7d8c03d3d3c09067a4147d0/src/cargo/core/compiler/compilation.rs#L256)), which must be a semver string, might I add, along with the other required compile-time values, I imagine that your application won't pass the build phase to begin with.

Either way, though, I think it's probably not a good use of our time to bikeshed the usage of `expect()` versus `unwrap()` here in this thread. 😄

Reply:
> The expect() message is useful at run-time when debugging broken run-time behavior, sure, but env!() is a compile-time invocation.

It is called on the `parse()`, not on `env!()`. 

If compiler has bug, then many things are possible. It could also build the project successfully, and it may  fail at the runtime. Like I said, _assertions_ are used to check your _assumptions_. Everytime it fails, you _learn_ new things, new possibility. An _assertion_, when fails, reveals extremely-hard-to-find-bugs somewhere in your tech stack, from your code to the compiler, even in the hardward.

Anyway, I'm okay with going forward and merge it, as I've said in my very first comment, even though we differ on how we think of _assertions_ and _assumptions_.

Reply:
The whole `expect()` vs `unwrap()` argument will become moot after https://github.com/rust-lang/rust/pull/67887 gets release in 1.42

Reply:
Wow. That's an awesome addition, very useful for developers. 

However, `expect` would still make sense because it adds custom message for _users_, not necessarily _developers_. Good messages are helpful for developers as well, though. But that is completely my opinion. :-)